lib/srv/db/dynamodb/engine.go

```diff
@@ -0,0 +1,421 @@
+	}
+	e.Log.Debugf("DynamoDB connection error: %v", err)
+
+	// try to convert to a trace err if we can.
+	code := trace.ErrorToCode(err)
+	body, err := json.Marshal(jsonErr{
+		Code:    strconv.Itoa(code),
+		Message: err.Error(),
+	})
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+	response := &http.Response{
+		Status:     http.StatusText(code),
+		StatusCode: code,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		// ContentLength is the authoritative value in the response,
+		// no need to also add the "Content-Length" header (source: go doc http.Response.Header).
+		ContentLength: int64(len(body)),
+		Header: map[string][]string{
+			"Content-Type": {"application/json"},
+		},
+		Body: ioutil.NopCloser(bytes.NewBuffer(body)),
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.WithError(err).Error("failed to send error response to DynamoDB client")
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target DynamoDB server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx, e.sessionCtx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// process reads request from connected dynamodb client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, req *http.Request) error {
+	defer req.Body.Close()
+	body, err := io.ReadAll(io.LimitReader(req.Body, teleport.MaxHTTPRequestSize))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	service, err := e.getService(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	signingName, err := serviceToSigningName(service)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	uri, err := e.getTargetURI(service)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	roundTripper, err := e.getRoundTripper(ctx, uri)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// rewrite the request URL and headers before signing it.
+	reqCopy, err := rewriteRequest(ctx, req, uri, bytes.NewReader(body))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	region := e.sessionCtx.Database.GetAWS().Region
+	roleArn := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, region, e.sessionCtx.Database.GetAWS().AccountID)
+	signedReq, err := e.signingSvc.SignRequest(reqCopy,
+		&libaws.SigningCtx{
+			SigningName:   signingName,
+			SigningRegion: region,
+			Expiry:        e.sessionCtx.Identity.Expires,
+			SessionName:   e.sessionCtx.Identity.Username,
+			AWSRoleArn:    roleArn,
```

Comment:
we do not.

@smallinsky we discussed adding this before and iirc decided to not add external id support in the initial implementation, unless I'm mistaken, but to be sure: should we support this? If we do then I think putting it in `db_service->databases[i].aws.external_id` would be the appropriate place.

Reply:
The sounds ok and handling `external_id`  looks like minor change so we can add it now. 
> If we do then I think putting it in db_service->databases[i].aws.external_id would be the appropriate place.

Yea, that we can att  it will be consistent with current app access approach where external_id is in handled by `apps[i].aws.external_id`